theano/scalar/basic.py

```diff
@@ -3605,30 +3610,56 @@ def __init__(self, inputs, outputs):
-                                        for var in outputs]):
+        if not any([isinstance(var.owner.op, Composite) for var in outputs]):
             # No inner Composite
             inputs, outputs = gof.graph.clone(inputs, outputs)
         else:
             # Inner Composite that we need to flatten
-            assert len(outputs) == 1
             # 1. Create a new graph from inputs up to the
             # Composite
+            out1 = []
+            for o in outputs:
+                for i in o.owner.inputs:
+                    out1.append(i)
             res = theano.compile.rebuild_collect_shared(
                 inputs=inputs,
-                outputs=outputs[0].owner.inputs,
+                outputs=out1,
                 copy_inputs_over=False)  # Clone also the inputs
-            # 2. We continue this partial clone with the graph in
-            # the inner Composite
-            res2 = theano.compile.rebuild_collect_shared(
-                inputs=outputs[0].owner.op.inputs,
-                outputs=outputs[0].owner.op.outputs,
-                replace=dict(izip(outputs[0].owner.op.inputs, res[1]))
-            )
-            assert len(res2[1]) == len(outputs)
+
+            # 2. We continue this partial clone with the graph in the
+            # inner Composite. If there is multiple outputs, o can be
+            # something else then a composite.
+            new_outputs = []
+            node_done = {}
+            for o_idx, o in enumerate(outputs):
+                if o.owner in node_done:
+                    out = node_done[o.owner]
+                elif not isinstance(o.owner.op, Composite):
+                    repl = {}  # replacement
+                    for i in o.owner.inputs:
+                        repl[i] = res[1][out1.index(i)]
+                    res2 = theano.compile.rebuild_collect_shared(
+                        inputs=o.owner.inputs,
+                        outputs=o.owner.outputs,
+                        replace=repl)
+                    out = res2[1]
+                    node_done[o.owner] = out
+                else:
+                    repl = {}  # replacement
+                    for i_extern, i_intern in zip(o.owner.inputs,
+                                                  o.owner.op.inputs):
+                        repl[i_intern] = res[1][out1.index(i_extern)]
+                    res2 = theano.compile.rebuild_collect_shared(
+                        inputs=o.owner.op.inputs,
+                        outputs=o.owner.op.outputs,
+                        replace=repl)
+                    out = res2[1]
+                    node_done[o.owner] = out
+                new_outputs.append(out[o.owner.outputs.index(o)])
+            assert len(new_outputs) == len(outputs)
             assert len(res[0]) == len(inputs)
             assert res[0] != inputs
-            inputs, outputs = res[0], res2[1]
+            inputs, outputs = res[0], new_outputs
```

Comment:
I don't understand. I tried to implement this case as a test and it pass:

```
+        # Test a corner case
+        z = theano.scalar.float32()
+        c = theano.scalar.Composite([x, y], [x+y])
+        cc = theano.scalar.Composite([x, y, z], [c(x, y), x+z])
+        Z = tensor.matrix()
+        o = tensor.Elemwise(scalar_op=cc)(X, Y, Z)
+
+        f = theano.function([X, Y, Z], o, mode=mode)
+        topo = f.maker.fgraph.toposort()
+        assert len(topo) == 1
+        assert len(topo[0].outputs) == 2
+        assert isinstance(topo[0].op, tensor.Elemwise)
+        assert isinstance(topo[0].op.scalar_op, theano.scalar.Composite)
+        utt.assert_allclose(f([[1.]], [[3.]], [[5.]])[0], [[4.]])
+        utt.assert_allclose(f([[1.]], [[3.]], [[5.]])[1], [[6.]])
```

Is this the case that you talk about?